### ISSUE TITLE
Fabric support?

### ISSUE THREAD

**Issue by anonyco** [JorelAli/CommandAPI#308]
### Description

Are there presently any plans for a Fabric port of the CommandAPI?, or what does the roadmap for this look like?

**Comment by JorelAli** [comment]
tl;dr: There are no existing plans for Fabric support for the CommandAPI in the near future (within the next year or so).

## Why not?

The CommandAPI was created for Bukkit (and Spigot and Paper) because Bukkit/Spigot/Paper do not currently have an API that lets you utilize Minecraft's command UI. The command UI in Minecraft uses Brigadier and the commands use NMS (`net.minecraft.server.....`) classes, such as `BlockPos`, `Level` and `Vec3` which are not accessible or usable in the Bukkit/Spigot/Paper API which have their own objects (e.g. `Location` and `World`).

The CommandAPI has to not only provide a front-end for Bukkit/Spigot/Paper objects (e.g. `Location` and `Player`), but it also has to provide a conversion from NMS objects to Bukkit/Spigot/Paper objects. This conversion step is not necessary for Fabric because you can use NMS objects without the need for conversion.

I recently (in the past few months) tried supporting Velocity, but it turns out that it's very likely that I'd have to rewrite the entire CommandAPI for velocity (i.e. there would be very little "shared common code" between the plugin and velocity version), so presumably porting to Fabric would be a giant task as well.

## Is there a possibility of Fabric support in the future?

Yes! Despite being primarily a plugin developer, I have poked at Fabric mods in the past and fabric opens the door for the one limitation of the CommandAPI, which is custom arguments, as briefly stated in the README:

> **'True' custom arguments and server-side argument implementations**
>
> Through some brief testing of the [regex-mod branch](https://github.com/JorelAli/CommandAPI/tree/dev/regex-mod) and my [MinecraftRegexArgumentMod](https://github.com/JorelAli/MinecraftRegexArgumentMod) repository, it was discovered that 'true' custom arguments (arguments with a custom implementation of the returned type and parser) are possible with the aid of a client-sided mod. Additionally, this functionality also works without a client-sided mod, assuming this is only used server-side. This can be useful for server-only datapacks, functions and tags, as well as command blocks.

At some point in the far future, I plan to add (or at the very least, officially support) some form of fabric support, either with the CommandAPI being a fabric client-side mod for use with the server plugin version of the CommandAPI, or as a fabric port (which would be a significant amount of work).

**Comment by anonyco** [comment]
Very good to know.

Thank you for the detailed explanation and I'm sure other people researching this topic will find it most useful.
